pkg/server/server.go

```diff
@@ -237,18 +244,55 @@ func New(dir string, listen string, enableTLS bool, diskMode bool, clientSession
 	kineConfig.Endpoint = fmt.Sprintf("dqlite://k8s?%s", params.Encode())
 
 	return &Server{
-		app:        app,
+		app: app,
+		swatcherConfig: StorageWatcherConfig{
+			storageDir:         dir,
+			storageWatchTicker: time.NewTicker(time.Duration(storageWatchPeriod) * time.Second),
+			storageWatchQuit:   make(chan struct{}),
+			diskThreshold:      storageDiskThreshold,
+		},
 		kineConfig: kineConfig,
 	}, nil
 }
 
+func (s *Server) WatchStorage() {
+	for {
+		select {
+		case <-s.swatcherConfig.storageWatchTicker.C:
+			isFull, err := IsStorageFull(s.swatcherConfig.storageDir, s.swatcherConfig.diskThreshold)
+			// If available space is less than threshold transfer roles and shutdown.
+			if err != nil {
+				logrus.WithError(err).Errorf("failed to check storage capacity")
+			} else if isFull {
+				// Create a separate context with 30 seconds to cleanup
+				stopCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+				defer cancel()
+				logrus.Debug("Handing over dqlite leadership")
+				if err := s.app.Handover(stopCtx); err != nil {
+					logrus.WithError(err).Errorf("Failed to handover dqlite")
+				}
+				logrus.Debug("Closing dqlite application")
+				if err := s.app.Close(); err != nil {
+					logrus.WithError(err).Errorf("failed to close dqlite app")
+				}
+				logrus.WithField("dir", s.swatcherConfig.storageDir).Fatalf("Disk is critically low, gracefully shutdown...")
+			}
+		case <-s.swatcherConfig.storageWatchQuit:
+			s.swatcherConfig.storageWatchTicker.Stop()
+			return
+		}
+	}
+}
```

Comment:
This should instead accept the `context.Context` passed in start, and return in `case <-ctx.Done()`